### ISSUE TITLE
Missing ops in scripted function

### ISSUE THREAD

**Issue by angelolovatto** [pytorch/pytorch#42411]
## 🐛 Bug

Hi everyone,

I’ve come across some interesting behavior regarding TorchScript and index ops. The following two functions do the same job.

## To Reproduce

```python
import torch
from torch import Tensor

@torch.jit.script
def fit(data: Tensor):
    # Reduce all but the last dimension
    reduction_dims = [i for i in range(data.dim() - 1)]

    data_mu = torch.mean(data, dim=reduction_dims, keepdim=True)
    data_sigma = torch.std(data, dim=reduction_dims, keepdim=True)
    data_sigma = torch.where(data_sigma < 1e-12, torch.ones_like(data_sigma), data_sigma)
    return data_mu, data_sigma

print(fit.code)

@torch.jit.script
def fit(data: Tensor):
    reduction_dims = [i for i in range(data.dim() - 1)]

    data_mu = torch.mean(data, dim=reduction_dims, keepdim=True)
    data_sigma = torch.std(data, dim=reduction_dims, keepdim=True)
    data_sigma[data_sigma < 1e-12].fill_(1.0)
    return data_mu, data_sigma

print(fit.code)
```

Outputs:
```bash
def fit(data: Tensor) -> Tuple[Tensor, Tensor]:
  reduction_dims = annotate(List[int], [])
  for i in range(torch.sub(torch.dim(data), 1)):
    _0 = torch.append(reduction_dims, i)
  data_mu = torch.mean(data, reduction_dims, True, dtype=None)
  data_sigma = torch.std(data, reduction_dims, True, True)
  _1 = torch.lt(data_sigma, 9.9999999999999998e-13)
  _2 = torch.ones_like(data_sigma, dtype=None, layout=None, device=None, pin_memory=None, memory_format=None)
  data_sigma0 = torch.where(_1, _2, data_sigma)
  return (data_mu, data_sigma0)

def fit(data: Tensor) -> Tuple[Tensor, Tensor]:
  reduction_dims = annotate(List[int], [])
  for i in range(torch.sub(torch.dim(data), 1)):
    _0 = torch.append(reduction_dims, i)
  data_mu = torch.mean(data, reduction_dims, True, dtype=None)
  data_sigma = torch.std(data, reduction_dims, True, True)
  return (data_mu, data_sigma)
```

## Expected behavior

Upon inspection of the TorchScript code, one can see that in the latter the JIT compiler completely removes an assignment operator: data_sigma[data_sigma < 1e-12].fill_(1.0)

Does anyone have a good explanation for this behavior? I’m worried now that the same may be happening in other parts of my code. Such assignments are important when avoiding numerical precision errors.

Cheers,
Ângelo

## Environment

PyTorch version: 1.5.1
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.15.6
GCC version: Could not collect
CMake version: version 3.17.3

Python version: 3.8
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.19.0
[pip3] torch==1.5.1
[conda] Could not collect

cc @suo @gmagogsfm

**Comment by ak-source** [comment]
```data_sigma[data_sigma < 1e-12]```
this creates a new tensor (there is no masked array concept in pytorch), so fill_ result is never used

**Comment by angelolovatto** [comment]
Indeed, I overlooked that fact. Here's the new example, which produces the expected output.

```python
@torch.jit.script
def fit(data: Tensor):
    reduction_dims = [i for i in range(data.dim() - 1)]

    data_mu = torch.mean(data, dim=reduction_dims, keepdim=True)
    data_sigma = torch.std(data, dim=reduction_dims, keepdim=True)
    data_sigma[data_sigma < 1e-12] = 1.0
    return data_mu, data_sigma

print(fit.code)
```

Output
```python
def fit(data: Tensor) -> Tuple[Tensor, Tensor]:
  reduction_dims = annotate(List[int], [])
  for i in range(torch.sub(torch.dim(data), 1)):
    _0 = torch.append(reduction_dims, i)
  data_mu = torch.mean(data, reduction_dims, True, dtype=None)
  data_sigma = torch.std(data, reduction_dims, True, True)
  _1 = torch.lt(data_sigma, 9.9999999999999998e-13)
  _2 = torch.tensor(1., dtype=ops.prim.dtype(data_sigma), device=ops.prim.device(data_sigma), requires_grad=False)
  _3 = annotate(List[Optional[Tensor]], [_1])
  _4 = torch.index_put_(data_sigma, _3, _2, False)
  return (data_mu, data_sigma)
```

Thank you so much for your help. I think it's safe to close this issue now.
